services-custom/dynamodb-enhanced/src/main/java/software/amazon/awssdk/enhanced/dynamodb/update/UpdateExpression.java

```diff
@@ -0,0 +1,234 @@
+        if (removeActions != null ? ! removeActions.equals(that.removeActions) : that.removeActions != null) {
+            return false;
+        }
+        if (setActions != null ? ! setActions.equals(that.setActions) : that.setActions != null) {
+            return false;
+        }
+        if (deleteActions != null ? ! deleteActions.equals(that.deleteActions) : that.deleteActions != null) {
+            return false;
+        }
+        return addActions != null ? addActions.equals(that.addActions) : that.addActions == null;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = removeActions != null ? removeActions.hashCode() : 0;
+        result = 31 * result + (setActions != null ? setActions.hashCode() : 0);
+        result = 31 * result + (deleteActions != null ? deleteActions.hashCode() : 0);
+        result = 31 * result + (addActions != null ? addActions.hashCode() : 0);
+        return result;
+    }
+
+    /**
+     * A builder for {@link UpdateExpression}
+     */
+    public static final class Builder {
+
+        private Collection<RemoveUpdateAction> removeActions;
+        private Collection<SetUpdateAction> setActions;
+        private Collection<DeleteUpdateAction> deleteActions;
+        private Collection<AddUpdateAction> addActions;
+
+        private Builder() {
+        }
+
+        /**
+         * Add an action of type {@link RemoveUpdateAction}
+         */
+        public Builder addAction(RemoveUpdateAction action) {
+            if (this.removeActions == null) {
+                this.removeActions = new ArrayList<>();
+            }
+            removeActions.add(action);
+            return this;
+        }
+
+        /**
+         * Add an action of type {@link SetUpdateAction}
+         */
+        public Builder addAction(SetUpdateAction action) {
+            if (this.setActions == null) {
+                this.setActions = new ArrayList<>();
+            }
+            setActions.add(action);
+            return this;
+        }
+
+        /**
+         * Add an action of type {@link DeleteUpdateAction}
+         */
+        public Builder addAction(DeleteUpdateAction action) {
+            if (this.deleteActions == null) {
+                this.deleteActions = new ArrayList<>();
+            }
+            deleteActions.add(action);
+            return this;
+        }
+
+        /**
+         * Add an action of type {@link AddUpdateAction}
+         */
+        public Builder addAction(AddUpdateAction action) {
+            if (this.addActions == null) {
+                this.addActions = new ArrayList<>();
+            }
+            addActions.add(action);
+            return this;
+        }
+
+        /**
+         * Adds a collection of {@link UpdateAction} of any subtype to the builder. Calling this operation
+         * repeatedly will add the new collection to existing items and not overwrite them.
+         */
+        public Builder actions(Collection<UpdateAction> actions) {
```

Comment:
? extends UpdateAction?

Reply:
Yes, I think order does matter, though the code that handles validations/creating final expressions hasn't been written yet, so that would make sense.

Reply:
@millems Why would the signature `? extends UpdateAction` be better in this case? We're not modifying the collection/list and in fact we do accept a mix of UpdateAction, so it's almost clearer to use UpdateAction.

Reply:
It would be so that a customer can pass in a `Collection<AddUpdateAction>`. They could still pass in a `Collection<UpdateAction>`. If we're read-only, it's proper to use covariant types.

Reply:
Ah, that's the use case I didn't try. Just feeding the subtypes into the method didn't trigger a difference.